pkg/defaults/defaults_global.go

```diff
@@ -0,0 +1,88 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package defaults
+
+import (
+	"os"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/defaults"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+func GetglobalBool(cmd *cobra.Command, key string) (bool, error) {
+	globalBoolMap := map[string]bool{
+		"debug":             false,
+		"debug-full":        false,
+		"insecure-registry": false,
+	}
+	if cmd.Flags().Changed(key) {
+		return cmd.Flags().GetBool(key)
+	}
+	vp := viper.GetViper()
+	configBool := vp.Get(key)
+	if configBool != nil {
+		return configBool.(bool), nil
+	} else {
+		if val, ok := globalBoolMap[key]; ok {
+			return val, nil
+		} else {
+			return false, nil
+		}
+	}
+}
+
+func GetglobalString(cmd *cobra.Command, key string) (string, error) {
+	globalStringMap := map[string]string{
+		"address":         defaults.DefaultAddress,
+		"namespace":       namespaces.Default,
+		"snapshotter":     containerd.DefaultSnapshotter,
+		"cni-path":        CNIPath(),
+		"cni-netconfpath": CNINetConfPath(),
+		"data-root":       DataRoot(),
+		"cgroup-manager":  CgroupManager(),
+	}
+
+	globalEnvMap := map[string]string{
+		"address":         "CONTAINERD_ADDRESS",
+		"namespace":       "CONTAINERD_NAMESPACE",
+		"snapshotter":     "CONTAINERD_SNAPSHOTTER",
+		"cni-path":        "CNI_PATH",
+		"cni-netconfpath": "NETCONFPATH",
+	}
+	if cmd.Flags().Changed(key) {
+		return cmd.Flags().GetString(key)
+	}
+	if envKey, ok := globalEnvMap[key]; ok {
+		envString := os.Getenv(envKey)
+		if envString != "" {
+			return envString, nil
+		}
+	}
```

Comment:
Can we fix `Changed()` to return `true` when EnvVar is set?

cc @robberphex

Reply:
> Leaving the results of some manual testing:
> 
> Some environment variables seem to have higher priority than command line options.
> 
> ```
> # CONTAINERD_ADDRESS=/tmp/dummy.sock nerdctl -a /run/containerd/containerd.sock version
> Client:
>  Version:	v0.11.1-164-g18f1151
>  Git commit:	18f1151de3fa320eda8dde4cf3403d3bde100e26
> FATA[0000] cannot access containerd socket "/tmp/dummy.sock": no such file or directory 
> ```
> 
> ```
> # CONTAINERD_SNAPSHOTTER=dummy nerdctl --snapshotter=overlayfs system info | grep "Storage Driver"
>  Storage Driver: dummy
> ```
> 
> `-H` doesn't seem to be enabled.
> 
> ```
> # nerdctl -H /tmp/nonexisting.sock version
> Client:
>  Version:	v0.11.1-164-g18f1151
>  Git commit:	18f1151de3fa320eda8dde4cf3403d3bde100e26
> 
> Server:
>  containerd:
>   Version:	v1.5.5
>   GitCommit:	72cec4be58a9eb6b2910f5d10f1c01ca47d231c0
> ```
> 
> `--storage-driver` doesn't seem to be enabled.
> 
> ```
> # nerdctl --storage-driver=dummy system info | grep "Storage Driver"
>  Storage Driver: overlayfs
> ```
> 
> cc @ningmingxiao @robberphex

maybe it is pflag bug see
https://github.com/spf13/pflag/pull/333

Reply:
> > > for `DEBUG` en var. suppose ther
> > > I mean
> > > rootCmd.PersistentFlags().Bool("debug", false, "debug mode")  not set  EnvVars:   []string{"DEBUG"},
> > > does viper.Get("debug")  still get  environment variable "DEBUG"?
> 
> https://pkg.go.dev/github.com/spf13/viper#Get Viper will check in the following order: override, flag, env, config file, key/value store, default
> 
> if you make `export DEBUG=true` in the machine this will override `rootCmd.PersistentFlags().Bool("debug", false, "debug mode")`

here is my example
main.go
```
package main

import (
	"fmt"
	"github.com/spf13/viper"
)
func main() {
	viper.SetConfigName("test.toml")
	viper.SetConfigType("toml")
	viper.AddConfigPath(".")
	err := viper.ReadInConfig()
	if err!=nil{
		fmt.Println(err)
	}
	fmt.Println(viper.Get("debug"))
	fmt.Println(viper.Get("DEBUG"))
}
```
cat test.toml
```
port = "3306"
```
env
```
root@nmx-System-Product-Name:~# env|grep -i debug
debug=true
DEBUG=true
```
run main.go
```
<nil>
<nil>
```